cartographer/mapping/sparse_pose_graph.cc

```diff
@@ -38,24 +36,52 @@ proto::SparsePoseGraph::Constraint::Tag ToProto(
   LOG(FATAL) << "Unsupported tag.";
 }
 
+std::unordered_map<const Submaps*, int> IndexTrajectories(
+    const std::vector<const Submaps*>& trajectories) {
+  std::unordered_map<const Submaps*, int> result;
+  for (const auto& trajectory : trajectories) {
+    result.emplace(trajectory, result.size());
+  }
+  return result;
+}
+
 void GroupTrajectoryNodes(
     const std::vector<TrajectoryNode>& trajectory_nodes,
+    const std::unordered_map<const Submaps*, int>& trajectory_indices,
     std::vector<std::vector<TrajectoryNode>>* grouped_nodes,
     std::vector<std::pair<int, int>>* new_indices) {
   CHECK_NOTNULL(grouped_nodes)->clear();
   CHECK_NOTNULL(new_indices)->clear();
 
-  std::unordered_map<const Submaps*, int> trajectory_ids;
-  for (const auto node : trajectory_nodes) {
-    const auto* trajectory = node.constant_data->trajectory;
-    if (trajectory_ids.emplace(trajectory, grouped_nodes->size()).second) {
-      new_indices->emplace_back(grouped_nodes->size(), 0);
-      grouped_nodes->push_back({node});
-    } else {
-      const int id = trajectory_ids[trajectory];
-      new_indices->emplace_back(id, (*grouped_nodes)[id].size());
-      (*grouped_nodes)[id].push_back(node);
-    }
+  grouped_nodes->resize(trajectory_indices.size());
+
+  for (const auto& node : trajectory_nodes) {
+    auto index_iterator =
+        trajectory_indices.find(node.constant_data->trajectory);
+    CHECK(index_iterator != trajectory_indices.end());
+    const int index = index_iterator->second;
+    new_indices->emplace_back(index, (*grouped_nodes)[index].size());
+    (*grouped_nodes)[index].push_back(node);
+  }
+}
+
+// TODO(macmason): This function is very nearly a copy of GroupTrajectoryNodes.
+// Consider refactoring them to share an implementation.
+void GroupSubmapStates(
+    const std::vector<SparsePoseGraph::SubmapState>& submap_states,
+    const std::unordered_map<const Submaps*, int>& trajectory_indices,
+    std::vector<std::vector<SparsePoseGraph::SubmapState>>* grouped_submaps,
```

Comment:
I think `grouped_submaps` is unused and should thus be removed. Replace by a local variable `std::vector<int>` to just count how many submaps have been added to a particular trajectory, i.e. what the next index would be?

This change could also be done for the nodes variant of the function, but I'm not feeling strongly there since it is used right now, although that use could be removed if we add all trajectories to the proto first and rearrange the nested loops to a single loop using `new_indices` to find the place where to add each node.